core/src/main/java/com/rudderstack/android/sdk/core/RudderDeviceModeManager.java

```diff
@@ -221,43 +226,58 @@ private void isDeviceModeFactoriesNotPresent() {
      *
      */
     private void replayMessageQueue() {
-        List<Integer> messageIds = new ArrayList<>();
-        List<String> messages = new ArrayList<>();
-        do {
-            dbPersistentManager.fetchDeviceModeEventsFromDb(messageIds, messages, MAX_FLUSH_QUEUE_SIZE);
-            RudderLogger.logDebug(String.format(Locale.US, "RudderDeviceModeManager: replayMessageQueue: replaying old messages with factories. Count: %d", messageIds.size()));
-            for (int i = 0; i < messageIds.size(); i++) {
-                try {
-                    RudderMessage message = gson.fromJson(messages.get(i), RudderMessage.class);
-                    makeFactoryDump(message, messageIds.get(i), true);
-                } catch (Exception e) {
-                    RudderLogger.logError(String.format(Locale.US, "RudderDeviceModeManager: replayMessageQueue: Exception in dumping message %s due to %s", messages.get(i), e.getMessage()));
+        synchronized (replayMessageQueueLock) {
+            List<Integer> messageIds = new ArrayList<>();
+            List<String> messages = new ArrayList<>();
+            do {
+                messages.clear();
+                messageIds.clear();
+                dbPersistentManager.fetchDeviceModeEventsFromDbWithLimitAndOffset(messageIds, messages, MAX_FLUSH_QUEUE_SIZE, offset);
```

Comment:
dbPersistenceManager uses a lock itself. No point in synchronizing this

Reply:
I've updated the synchronisation logic, please review it again.